pkg/plugins/gateway/prefixcacheindexer/tree.go

```diff
@@ -0,0 +1,476 @@
+	isLeaf        bool
+	contextLength int
+	depth         int
+	ModelToPods   map[string]map[string]time.Time // model -> {podName -> lastAccessTime}
+}
+
+func (n *TreeNode) GetKey() []int {
+	return n.key
+}
+
+func (n *TreeNode) GetValue() []int {
+	return n.value
+}
+
+func (n *TreeNode) NumTokens() int {
+	return len(n.value)
+}
+
+func (n *TreeNode) ContextLength() int {
+	return n.contextLength
+}
+
+func (c *LPRadixCache) NewTreeNode(numPods int, parent *TreeNode, key []int, value []int) *TreeNode {
+	// Create the node with initialized maps and slices
+	node := &TreeNode{
+		id:            c.nextNodeID,
+		children:      make(map[int]*TreeNode),
+		parent:        parent,
+		key:           make([]int, len(key)),   // Allocate space for key
+		value:         make([]int, len(value)), // Allocate space for value (using len(value), not len(key))
+		refCounter:    make([]int, numPods),
+		load:          1,
+		lastAccess:    time.Now(),
+		evictedPods:   make(map[int]bool),
+		cachedPods:    make(map[int]bool),
+		ModelToPods:   make(map[string]map[string]time.Time),
+		depth:         0,
+		contextLength: 0,
+	}
+
+	// Increment node ID for next creation
+	klog.Infof("Created a new node(%d) with key: %v and value: %v", node.id, key, value)
+	c.nextNodeID++
+
+	// Set depth and context length based on parent
+	if parent != nil {
+		node.depth = parent.depth + 1
+		node.contextLength = parent.contextLength + len(key)
+	}
+
+	// Copy key and value slices
+	if len(key) > 0 {
+		copy(node.key, key)
+	}
+	if len(value) > 0 {
+		copy(node.value, value)
+	}
+
+	return node
+}
+
+func (c *LPRadixCache) PrettyPrint() {
+	// c.mu.RLock()
+	// defer c.mu.RUnlock()
+	c.prettyPrintHelper(c.rootNode, "", true)
+}
+
+func (c *LPRadixCache) prettyPrintHelper(node *TreeNode, prefix string, isLast bool) {
+	if node == nil {
+		return
+	}
+	marker := "└── "
+	if !isLast {
+		marker = "├── "
+	}
+	childPrefix := prefix + "    "
+	if !isLast {
+		childPrefix = prefix + "│   "
+	}
+	klog.Infof("%s%s[Node: %d, Key: '%v', Load: %d, Depth: %d]", prefix, marker, node.id, node.key, node.load, node.depth)
+	if len(node.ModelToPods) > 0 {
+		klog.Infof("%s    Models:", prefix)
+		for model, pods := range node.ModelToPods {
+			podNames := make([]string, 0, len(pods))
+			for podName := range pods {
+				podNames = append(podNames, podName)
+			}
+			klog.Infof("%s    └── %s: %v", prefix, model, podNames)
+		}
+	}
+	childKeys := make([]int, 0, len(node.children))
+	for k := range node.children {
+		childKeys = append(childKeys, k)
+	}
+	sort.Ints(childKeys)
+
+	for i, key := range childKeys {
+		isLastChild := i == len(childKeys)-1
+		c.prettyPrintHelper(node.children[key], childPrefix, isLastChild)
+	}
+}
+
+type LPRadixCache struct {
+	mu            sync.RWMutex
+	rootNode      *TreeNode
+	numPods       int
+	allocatedSize []int
+	allNodes      map[int]*TreeNode
+	nextNodeID    int
+	startTime     time.Time
+}
+
+func NewLPRadixCache(numPods int) *LPRadixCache {
```

Comment:
I didn't get the part why `numPods` is an argument for initialization? this would be changed dynamically in production cases

Reply:
We supposed to watch all the pods belonging to a model etc. a cache client could be embedded